### PR TITLE
deps: upgrade to typescript 4.x

### LINKED ISSUE #1441: [Engineering] TypeScript 3.7 migration plan
## Versions
What TypeScript version are you using? **3.5.3**

TypeScript 3.7 contains breaking changes when compared to 3.5. Our last two minor releases (4.5, 4.6) were shipped with Typescript 3.5.

TypeScript 3.7 has several [new features](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html) that could help make the team's lives easier, especially for those coming from C# and writing in TypeScript.

If we use the `.d.ts` files from TypeScript 3.7, it contains a breaking change regarding the declaration of get/set accessors in an ambient context.

As such, we can explore using different packages to generate `.d.ts` files if we as a dev team want to move to TS 3.7, but keep our current types. One package suggested by @christopheranderson is [`@microsoft/api-extractor`](https://www.npmjs.com/package/@microsoft/api-extractor) ([GitHub Repo](https://github.com/microsoft/rushstack/tree/master/apps/api-extractor))

___

Additional context: https://github.com/microsoft/botbuilder-js/issues/1436

### PR DESCRIPTION
**Pull request by joshgummersall**
Fixes #1441

Note: it looks like our use of `antlr` doesn't play nicely with the newer Typescript compiler. In a few places, `skipLibCheck` fixes the compilation errors. In others, unfortunately, it does not. I believe `adaptive-expressions` and its derivatives were the only examples where `typescript@3.5.3` is required.

### CODE REVIEW

File: libraries/botbuilder-dialogs-adaptive/src/actions/switchCondition.ts
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/actions/switchCondition.ts:35]: I think this actually more accurately describes the converter. This fixes a compilation error as well.

**Comment by Danieladu** [comment]
hi @joshgummersall  About the version of antlr target. I have an idea that to use [antlr4](https://www.npmjs.com/search?q=antlr4) and [@types/antlr4](https://www.npmjs.com/package/@types/antlr4) instead of the original [antlr4ts](https://www.npmjs.com/package/antlr4ts) which is not maintained in the future anymore.
The only concern is we should add the type for lexer/visitor/parser js file manually. Or could we make the "allowJs" true for the typescript build? 
Looking forward to your reply.

**Comment by joshgummersall** [comment]
I'm hesitant to move to allowJs. I'm also not familiar with that package so I'll have to investigate. Are the types you mentioned not expressed in @types/antlr4?

**Comment by Danieladu** [comment]
> I'm hesitant to move to allowJs. I'm also not familiar with that package so I'll have to investigate. Are the types you mentioned not expressed in @types/antlr4?

The @types/antlr4 just provide the types of the antlr SDK library without the types of the generated files(lexer/parser/listener/visitor).

**Comment by joshgummersall** [comment]
Got it. I'll investigate a little more tomorrow. Maybe we can figure out a reasonable way to generate some typings.

**Comment by stevengum** [comment]
Speaking of typings, when this PR is ready to merge, let's cut a daily build prior to merge and ask the Southworks team working with Solutions to test the SDK built with `TypeScript 4.x`.

They caught the typings issue last year due to different TypeScript versions being used for our SDK.

**Comment by joshgummersall** [comment]
@stevengum I've cut a build of this (`4.12.0-ts4-09c112b1af10`) and tested it locally with every Typescript sample. They all compiled successfully. I'm hesitant to block merging this PR right now since all our tests (including the consumer test) are passing.

I do think we should cut a build once it lands and have Southworks test it, especially if they are doing more extensive testing than what I described. In the case that they are testing more extensively we should also codify that and run it as some kind of CI job.

**Review by stevengum — APPROVED** [review]

**Comment by joshgummersall** [comment]
Note: rebased to clean up commit log. No code changes.

**Comment by Batta32** [comment]
Hi @stevengum & @joshgummersall, we successfully validated the build `4.12.0-ts4-09c112b1af10` and confirmed that it's correctly working with the TypeScript Solutions components!

**Pull request merged**
